src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs

```diff
@@ -595,6 +596,32 @@ internal bool IsEntryPointCandidate
             get { return IsStatic && Name == WellKnownMemberNames.EntryPointMethodName; }
         }
 
+
+        internal bool ReturnsAwaitableToVoidOrInt(CSharpCompilation compilation)
+        {
+            // Early bail so we only even check things that are System.Threading.Tasks.Task(<T>)
+            if (ReturnType is NamedTypeSymbol namedType &&
+                !(namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task) ||
+                  namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T)))
+            {
+                return false;
+            }
+
+
+            var dumbInstance = new BoundLiteral(CSharpSyntaxTree.Dummy.GetRoot(), ConstantValue.Null, ReturnType);
+            var syntax = this.ExtractReturnTypeSyntax();
+            var binder = compilation.GetBinder(syntax);
+            BoundExpression result;
+            var bag = new DiagnosticBag();
+            var success = binder.GetAwaitableExpressionInfo(dumbInstance, out _, out _, out _, out result, syntax, bag, false, false);
+
+            return
+                success &&
+                bag.IsEmptyWithoutResolution &&
```

Comment:
>bag.IsEmptyWithoutResolution && [](start = 16, length = 31)

We would want to check for errors here (without resolution), warnings and suggestions are fine.